bigquery/storage/managedwriter/adapt/protoconversion.go

```diff
@@ -0,0 +1,259 @@
+	storagepb.TableFieldSchema_REQUIRED: descriptorpb.FieldDescriptorProto_LABEL_OPTIONAL,
+}
+
+// Allows conversion between BQ schema type and FieldDescriptorProto's type.
+var bqTypeToFieldTypeMap = map[storagepb.TableFieldSchema_Type]descriptorpb.FieldDescriptorProto_Type{
+	storagepb.TableFieldSchema_BIGNUMERIC: descriptorpb.FieldDescriptorProto_TYPE_BYTES,
+	storagepb.TableFieldSchema_BOOL:       descriptorpb.FieldDescriptorProto_TYPE_BOOL,
+	storagepb.TableFieldSchema_BYTES:      descriptorpb.FieldDescriptorProto_TYPE_BYTES,
+	storagepb.TableFieldSchema_DATE:       descriptorpb.FieldDescriptorProto_TYPE_INT32,
+	storagepb.TableFieldSchema_DATETIME:   descriptorpb.FieldDescriptorProto_TYPE_INT64,
+	storagepb.TableFieldSchema_DOUBLE:     descriptorpb.FieldDescriptorProto_TYPE_DOUBLE,
+	storagepb.TableFieldSchema_GEOGRAPHY:  descriptorpb.FieldDescriptorProto_TYPE_STRING,
+	storagepb.TableFieldSchema_INT64:      descriptorpb.FieldDescriptorProto_TYPE_INT64,
+	storagepb.TableFieldSchema_NUMERIC:    descriptorpb.FieldDescriptorProto_TYPE_BYTES,
+	storagepb.TableFieldSchema_STRING:     descriptorpb.FieldDescriptorProto_TYPE_STRING,
+	storagepb.TableFieldSchema_STRUCT:     descriptorpb.FieldDescriptorProto_TYPE_MESSAGE,
+	storagepb.TableFieldSchema_TIME:       descriptorpb.FieldDescriptorProto_TYPE_INT64,
+	storagepb.TableFieldSchema_TIMESTAMP:  descriptorpb.FieldDescriptorProto_TYPE_INT64,
+}
+
+// For TableFieldSchema OPTIONAL mode, we use the wrapper types to allow for the
+// proper representation of NULL values, as proto3 semantics would just use default value.
+var bqTypeToWrapperMap = map[storagepb.TableFieldSchema_Type]string{
+	storagepb.TableFieldSchema_BIGNUMERIC: ".google.protobuf.BytesValue",
+	storagepb.TableFieldSchema_BOOL:       ".google.protobuf.BoolValue",
+	storagepb.TableFieldSchema_BYTES:      ".google.protobuf.BytesValue",
+	storagepb.TableFieldSchema_DATE:       ".google.protobuf.Int32Value",
+	storagepb.TableFieldSchema_DATETIME:   ".google.protobuf.Int64Value",
+	storagepb.TableFieldSchema_DOUBLE:     ".google.protobuf.DoubleValue",
+	storagepb.TableFieldSchema_GEOGRAPHY:  ".google.protobuf.StringValue",
+	storagepb.TableFieldSchema_INT64:      ".google.protobuf.Int64Value",
+	storagepb.TableFieldSchema_NUMERIC:    ".google.protobuf.BytesValue",
+	storagepb.TableFieldSchema_STRING:     ".google.protobuf.StringValue",
+	storagepb.TableFieldSchema_TIME:       ".google.protobuf.Int64Value",
+	storagepb.TableFieldSchema_TIMESTAMP:  ".google.protobuf.Int64Value",
+}
+
+// filename used by well known types proto
+var wellKnownTypesWrapperName = "google/protobuf/wrappers.proto"
+
+// dependencyCache is used to reduce the number of unique messages we generate by caching based on the tableschema.
+//
+// keys are based on the base64-encoded serialized tableschema value.
+type dependencyCache map[string]protoreflect.Descriptor
+
+func (dm *dependencyCache) get(schema *storagepb.TableSchema) protoreflect.Descriptor {
```

Comment:
nit: does not need to be a pointer as a map is already.

Reply:
Blerg, thanks for the spot.  Switched.